### PR TITLE
Flink: refactor DataIterator to use composition (instead of inheritance)

### PR DESCRIPTION
**Pull request by stevenzwu**
With this composition mode, `DataIterator` deals with `CombinedScanTask` and `IteratorReader` deals with individual `FileScanTask`. `DataIterator` use composition to reference `IteratorReader`

Here are the motivations behind this change
1) address @openinx 's [review comment](https://github.com/apache/iceberg/pull/2105#discussion_r568303231)
2) make it easier to extend. E.g., internally we would like to add another converter from RowData to some other type Foo. We can't define some `FooDataIterator extends RowDataIterator` as the generic type is already fixed to `RowData`. With this new composition model, we can define a `FooIteratorReader implements IteratorReader` and `FooIteratorReader` can use `RowDataIteratorReader` by composition.

### CODE REVIEW

**Comment by stevenzwu** [comment]
@openinx @JingsongLi @rdblue can you help take a look?

File: flink/src/main/java/org/apache/iceberg/flink/source/DecryptedInputFiles.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.encryption.EncryptedFiles;
+import org.apache.iceberg.encryption.EncryptedInputFile;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+class DecryptedInputFiles {
```
Comment (rdblue) [inline flink/src/main/java/org/apache/iceberg/flink/source/DecryptedInputFiles.java:36]: I think this class is generally useful and should probably go in core rather than in Flink. That would allow us to deduplicate the code between here and Spark.

Also, what about calling this `InputFileDecryptor`? The current name seems to imply that this is a set of input files and I wouldn't expect the API that is exposed.

File: flink/src/main/java/org/apache/iceberg/flink/source/IteratorReader.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+package org.apache.iceberg.flink.source;
+
+import java.io.Serializable;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.io.CloseableIterator;
+
+/**
+ * Read a {@link FileScanTask} into a {@link CloseableIterator}
+ */
+public interface IteratorReader<T> extends Serializable {
```
Comment (rdblue) [inline flink/src/main/java/org/apache/iceberg/flink/source/IteratorReader.java:29]: The name `IteratorReader` doesn't make sense to me because iterators aren't generally read. They are iterated through. It also seems strange to me to have a method that is essentially a way to create an iterator but is not a `CloseableIterable`. Is there a way to restructure this so that this is a `CloseableIterable` instead?

Last, why does this need to be `Serializable`?

File: flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java [hunk 3]
```diff
@@ -106,12 +77,15 @@ private void updateCurrentIterator() {
...
+    return iteratorReader.open(scanTask, decryptedInputFiles);
+  }
 
   @Override
   public void close() throws IOException {
     // close the current iterator
     currentIterator.close();
     tasks = null;
   }
+
```
Comment (rdblue) [inline flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java:90]: Nit: unnecessary whitespace change

File: flink/src/main/java/org/apache/iceberg/flink/source/DecryptedInputFiles.java (hunk 1 above)
Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/DecryptedInputFiles.java:36]: sure. I can move it to the core and update Spark or other related places.

I will also rename it to `InputFilesDecryptor`

File: flink/src/main/java/org/apache/iceberg/flink/source/IteratorReader.java (hunk 2 above)
Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/IteratorReader.java:29]: what about `FileReader` as class name and change the method name from `open` to `read`? Its purpose is to read a `FileScanTask` into a `CloseableIterator` or `CloseableIterable`.

It needs to be `Serializable` because it gets shipped/serialized from jobmanager to taksmanager during deployment.

File: flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java (hunk 3 above)
Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java]: fixed

File: flink/src/main/java/org/apache/iceberg/flink/source/DecryptedInputFiles.java (hunk 1 above)
Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/DecryptedInputFiles.java]: renamed the class and moved it to the core module. I can update other places (like Spark in a follow-up PR).

File: flink/src/main/java/org/apache/iceberg/flink/source/IteratorReader.java (hunk 2 above)
Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/IteratorReader.java]: renamed it to `FileReader`

File: core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+public class InputFilesDecryptor {
+
+  private final Map<String, InputFile> decryptedInputFiles;
+
+  public InputFilesDecryptor(CombinedScanTask combinedTask, FileIO io, EncryptionManager encryption) {
```
Comment (openinx) [inline core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java:37]: Moving this into a common `InputFilesDecryptor` looks good to me, I think we could propose the following PR to make hive/spark integration to reuse it.

File: core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    combinedTask.files().stream()
+        .flatMap(fileScanTask -> Stream.concat(Stream.of(fileScanTask.file()), fileScanTask.deletes().stream()))
+        .forEach(file -> keyMetadata.put(file.path().toString(), file.keyMetadata()));
+    Stream<EncryptedInputFile> encrypted = keyMetadata.entrySet().stream()
+        .map(entry -> EncryptedFiles.encryptedInput(io.newInputFile(entry.getKey()), entry.getValue()));
+
+    // decrypt with the batch call to avoid multiple RPCs to a key server, if possible
+    Iterable<InputFile> decryptedFiles = encryption.decrypt(encrypted::iterator);
+
+    Map<String, InputFile> files = Maps.newHashMapWithExpectedSize(combinedTask.files().size());
```
Comment (openinx) [inline core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java:48]: The correct expected hash map size should be: `keyMetadata.size()`, because a `combinedTask` will have many files inside it , include both insert data files and delete files.

File: core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    decryptedFiles.forEach(decrypted -> files.putIfAbsent(decrypted.location(), decrypted));
+    this.decryptedInputFiles = Collections.unmodifiableMap(files);
+  }
+
+  public InputFile getInputFile(FileScanTask task) {
+    Preconditions.checkArgument(!task.isDataTask(), "Invalid task type");
```
Comment (openinx) [inline core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java:54]: Nit: I don't quite understand why we need an extra getInputFile for the `FileScanTask` with dataTask enabled.

File: flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java [hunk 7]
```diff
@@ -21,64 +21,36 @@
...
-import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
-import org.apache.iceberg.relocated.com.google.common.collect.Maps;
 
 /**
  * Base class of Flink iterators.
  *
  * @param <T> is the Java class returned by this iterator whose objects contain one or more rows.
  */
-abstract class DataIterator<T> implements CloseableIterator<T> {
+public class DataIterator<T> implements CloseableIterator<T> {
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java:37]: Why make this internal class  public ?

File: flink/src/main/java/org/apache/iceberg/flink/source/FileReader.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.io.Serializable;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.encryption.InputFilesDecryptor;
+import org.apache.iceberg.io.CloseableIterator;
+
+/**
+ * Read a {@link FileScanTask} into a {@link CloseableIterator}
+ */
+public interface FileReader<T> extends Serializable {
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/source/FileReader.java:30]: The name `FileReader` implies we are opening a given file and trying to iterate the rows from it.  But in fact,  we are trying to opening the bunch files of FileScanTask ( including one data file and several delete files maybe) and iterate the rows from them.   I will suggest to use `FileScanTaskReader`, that's more clear.

Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/FileReader.java:30]: +1. will change

File: flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java [hunk 9]
```diff
@@ -106,7 +78,9 @@ private void updateCurrentIterator() {
     }
   }
 
-  abstract CloseableIterator<T> openTaskIterator(FileScanTask scanTask) throws IOException;
+  private CloseableIterator<T> openTaskIterator(FileScanTask scanTask) throws IOException {
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java:81]: The throwing IOException could be removed now because we are throwing CheckedIOException, right ?

File: flink/src/main/java/org/apache/iceberg/flink/source/FileReader.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+  CloseableIterator<T> open(FileScanTask fileScanTask, InputFilesDecryptor inputFilesDecryptor);
+
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/source/FileReader.java:33]: Extra empty line ?

File: flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java (hunk 7 above)
Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java:37]: This is made public because flip-27 source code will need to construct it from a diff package like `org.apache.iceberg.flink.source.reader`.  I can add a `@Internal` annotation to those classes.

File: core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java (hunk 5 above)
Comment (stevenzwu) [reply core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java:48]: you are right. will update.

File: core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java (hunk 4 above)
Comment (stevenzwu) [reply core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java:37]: yeah. I can follow up with hive/spark integration in a follow-up PR

File: core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java (hunk 6 above)
Comment (stevenzwu) [reply core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java:54]: this method is called from readers (Avro, Parquet, Orc)
```
Parquet.ReadBuilder builder = Parquet.read(decryptedInputFiles.getInputFile(task))
        .reuseContainers()
```

The other `getInputFile(String location)` method is called from `FlinkDeleteFilter`

File: flink/src/main/java/org/apache/iceberg/flink/source/FileReader.java (hunk 10 above)
Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/FileReader.java:33]: sure. I have one empty line before and after the method. will remove both. It seems that Iceberg coding style doesn't use empty lines, e.g. `interface Listener`. 

I wish Iceberg has adopted formatting rules (like google-java-format) via spotless.

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java [hunk 11]
```diff
@@ -42,21 +42,22 @@
...
   private transient long currentReadCount = 0L;
 
   FlinkInputFormat(TableLoader tableLoader, Schema tableSchema, FileIO io, EncryptionManager encryption,
                    ScanContext context) {
     this.tableLoader = tableLoader;
-    this.tableSchema = tableSchema;
     this.io = io;
     this.encryption = encryption;
     this.context = context;
+    this.rowDataReader = new RowDataFileReader(tableSchema,
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java:59]: Is there necessary to expose the `RowDataFileReader` to the FlinkInputFormat ?  The calling chain is actually:  FlinkInputFormat ( whole scan) ->  DataIterator (CombinedScanTask reader) -> RowDataFileReader ( FileScanTask reader ).   I mean the FlinkInputFormat won't call the FileScanTask reader directly,  so  I'd prefer to hidden the internal details about how to construct the RowDataFileReader inside DataIterator. This is more in line with the interface design of software design.

File: flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileReader.java [hunk 12]
```diff
@@ -47,51 +45,51 @@
 import org.apache.iceberg.types.TypeUtil;
 import org.apache.iceberg.util.PartitionUtil;
 
-class RowDataIterator extends DataIterator<RowData> {
+public class RowDataFileReader implements FileReader<RowData> {
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileReader.java:48]: Nit: Better to add a comment indicating that the RowDataFileReader is designed for scanning rows from FileScanTask. Because as a reviewer, I cannot see the meaning directly from the class naming ( I think it's a historical issue in the initial review :-) )

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java (hunk 11 above)
Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java:59]: `DataIterator` is a generic type. The actual `FileScanTaskReader` needs to be provided when constructing `DataIterator`.  Previously, `FlinkInputFormat` constructs the extended class `RowDataIterator`, which is removed in the PR. Now `FlinkInputFormat` constructs the `RowDataFileScanTaskReader` and pass it into `DataIterator` constructor for composition.

File: flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileReader.java (hunk 12 above)
Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileReader.java:48]: there is Javadoc in the interface `FileScanTaskReader`. I think that is the right place for the Javadoc.

File: core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java (hunk 6 above)
Comment (openinx) [reply core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java:54]: Okay, it's used for reading the row  from `BaseFileScanTask` only,  don't allow to read the `DataTask`.   That's OK.

File: flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java [hunk 13]
```diff
@@ -21,64 +21,38 @@
...
 
 /**
- * Base class of Flink iterators.
+ * Flink data iterator that reads {@link CombinedScanTask} into a {@link CloseableIterator}
  *
- * @param <T> is the Java class returned by this iterator whose objects contain one or more rows.
+ * @param <T> is the output data type returned by this iterator.
  */
-abstract class DataIterator<T> implements CloseableIterator<T> {
+@Internal
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java:38]: The `Internal` annotation comes from flink,  in theory we should use the iceberg's  annotation which we still don't have.

File: core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.Map;
+import java.util.stream.Stream;
+import org.apache.iceberg.CombinedScanTask;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+public class InputFilesDecryptor {
```
Comment (openinx) [inline core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java:33]: Nit:  It's `InputFilesDecryptor` or `InputFilesDecrypter` ?

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java (hunk 11 above)
Comment (openinx) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java]: You mean there will be other data type applied to the generic `DataIterator` ( For example for flip-24 ) ?  There seems no other `FileScanTaskReader` implementation except the RowDataFileScanTaskReader.  I think we may don't want to the generic type for the `DataIterator`,  then we also don't need the extra `FileScanTaskReader` (we could just rename the RowDataFileScanTaskReader to FileScanTaskReader directly), and finally we construct the FileScanTaskReader inside the DataIterator.   How do you think ?

File: flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileScanTaskReader.java [hunk 15]
```diff
@@ -47,51 +46,52 @@
...
-    return iterable.iterator();
+    FlinkDeleteFilter deletes = new FlinkDeleteFilter(task, tableSchema, projectedSchema, inputFilesDecryptor);
+    return deletes
+        .filter(newIterable(task, deletes.requiredSchema(), idToConstant, inputFilesDecryptor))
+        .iterator();
   }
 
-  private CloseableIterable<RowData> newIterable(FileScanTask task, Schema schema, Map<Integer, ?> idToConstant) {
+  private CloseableIterable<RowData> newIterable(
+      FileScanTask task, Schema schema, Map<Integer, ?> idToConstant, InputFilesDecryptor inputFilesDecryptor) {
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileScanTaskReader.java:79]: Nit:  inputFilesDecryptor -> inputFilesDecrypter ?

File: flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileScanTaskReader.java [hunk 16]
```diff
@@ -47,51 +46,52 @@
...
+                                   String nameMapping, boolean caseSensitive) {
     this.tableSchema = tableSchema;
     this.projectedSchema = projectedSchema;
     this.nameMapping = nameMapping;
     this.caseSensitive = caseSensitive;
   }
 
   @Override
-  protected CloseableIterator<RowData> openTaskIterator(FileScanTask task) {
+  public CloseableIterator<RowData> open(FileScanTask task, InputFilesDecryptor inputFilesDecryptor) {
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileScanTaskReader.java:66]: If we construct the `RowDataFileScanTaskReader` inside the [DataIterator](https://github.com/apache/iceberg/pull/2905/files#diff-696823dda1232c713e658befb32b1c385a48b46c79ef25fc09edb70f5cc962a9R51),  then we don't need to pass the `InputFilesDecryptor`  for every `open` FileScanTask method, we also don't need to pass the `InputFilesDecryptor`  to the newXXXIterable methods (I mean those changes could be reverted) because they could just use the class' private `InputFilesDecryptor` instance.

**Review by openinx — COMMENTED** [review]
Thanks @stevenzwu for the refactoring,  this PR almost looks good to me.  Just left several comments.

File: core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java (hunk 14 above)
Comment (stevenzwu) [reply core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java:33]: Both Decryptor and Decrypter are valid. According to google, Decryptor is slightly more common (probably in American). Decrypter is probably more common in British English.

https://books.google.com/ngrams/graph?content=decryptor%2Cdecrypter&year_start=1800&year_end=2000&corpus=15&smoothing=3&share=&direct_url=t1%3B%2Cdecryptor%3B%2Cc0%3B.t1%3B%2Cdecrypter%3B%2Cc0

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java (hunk 11 above)
Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java]: It is correct that in iceberg-flink, there is no other impl of `FileScanTaskReader` other than `RowDataFileScanTaskReader`. But At Netflix and Apple, we use Avro schema for the Kafka data and deserialize bytes into Avro Record. For Iceberg source, we also want to deserialize data into Avro Record. Hence we will plug in `AvroGenericRecordFileScanTaskReader`. That is one of the motivation of this refactoring.

File: flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java (hunk 13 above)
Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java:38]: we can add an `@Internal` annotation to the `org.apache.iceberg.annotation` package in `iceberg-api`

File: flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileScanTaskReader.java (hunk 16 above)
Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileScanTaskReader.java:66]: This is also tied to the other comment where we need `DataIterator` to be generic.

What you described above is similar to the current status, where we have `RowDataIterator` extends from `DataIterator`. Due to inheritance, `RowDataIterator` can call protected methods/variables from base class.

If we switch to composition model, we need to pass in `InputFilesDecryptor` to the `RowDataFileScanTaskReader`

File: core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java (hunk 14 above)
Comment (openinx) [reply core/src/main/java/org/apache/iceberg/encryption/InputFilesDecryptor.java:33]: Okay,  thanks for the verification !

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java (hunk 11 above)
Comment (openinx) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java]: That make sense for providing an abstracted `FIleScanTaskReader` in the offical apache iceberg repo,  what concerns me is:  we may get this interface refactored  or removed when other developers propose a new pull request ,  if we don't have other impl in apache repo.  Will we have other impl for the flip-27 work in offical repo ?

Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java]: @openinx I extracted `AbstractFileScanTaskReader` as you suggested. 

In flip-27 source in apache iceberg repo, we won't have other `FileScanTaskReader` impl.

Comment (openinx) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java]: Sorry, I think I did not describe this more clear.  I mean the newly introduced `FileScanTaskReader` interface is nice to have in apache repo, so that others could provides  their own implementation, such as `AvroGenericRecordFileScanTaskReader`.   I don't think the latest commit's `AbstractFileScanTaskReader`  will help to share common code between the `AvroGenericRecordFileScanTaskReader`  & `RowDataFileScanTaskReader`  because its exposed classes such as `DeleteFilter`, `newAvroIterator`, `InputFilesDecryptor` are not stable enough, we may change/refactor them in the following PRs (The DeleteFilter will need a big refactor for supporting unified v2 compaction).   

Let's just revert the commit https://github.com/apache/iceberg/pull/2905/commits/72348c7898c42566a8a5a071426b32f8ea746cfd,  all the other things look great to me now,  I plan to merge this PR.  Thanks for the great work @stevenzwu  !

Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java]: @openinx sorry for the misunderstanding. I have reverted the last commit by hard reset and force push

**Review by openinx — APPROVED** [review]
LGTM
